Content.Server/Discord/DiscordLink/DiscordLink.cs

```diff
@@ -0,0 +1,255 @@
+    public string[] Arguments { get; init; } = Array.Empty<string>();
+    /// <summary>
+    /// Information about the message that the command was received from. This includes the message content, author, etc.
+    /// Use this to reply to the message, delete it, etc.
+    /// </summary>
+    public SocketMessage Message { get; init; } = default!;
+}
+
+public sealed class DiscordLink : IPostInjectInit
+{
+    [Dependency] private readonly ILogManager _logManager = default!;
+    [Dependency] private readonly IConfigurationManager _configuration = default!;
+    [Dependency] private readonly IReflectionManager _reflectionManager = default!;
+
+    /// <summary>
+    ///    The Discord client. This is null if the bot is not connected.
+    /// </summary>
+    /// <remarks>
+    ///     This should not be used directly outside of DiscordLink. So please do not make it public. Use the methods in this class instead.
+    /// </remarks>
+    private DiscordSocketClient? _client;
+    private ISawmill _sawmill = default!;
+    private ISawmill _sawmillNet = default!;
+
+    private ulong _guildId;
+    private string _botToken = string.Empty;
+
+    private bool _registeredLinks = false;
+
+    public string BotPrefix = default!;
+    /// <summary>
+    /// If the bot is currently connected to Discord.
+    /// </summary>
+    public bool IsConnected => _client != null;
+
+    #region Events
+
+    /// <summary>
+    ///     Event that is raised when a command is received from Discord.
+    /// </summary>
+    public event Action<CommandReceivedEventArgs>? OnCommandReceived;
+    /// <summary>
+    ///     Event that is raised when a message is received from Discord. This is raised for every message, including commands.
+    /// </summary>
+    public event Action<SocketMessage>? OnMessageReceived;
+
+    public void RegisterCommandCallback(Action<CommandReceivedEventArgs> callback, string command)
+    {
+        OnCommandReceived += args =>
+        {
+            if (args.Command == command)
+                callback(args);
+        };
+    }
+
+    #endregion
+
+    public void Initialize()
+    {
+        _configuration.OnValueChanged(CCVars.DiscordGuildId, OnGuildIdChanged, true);
+        _configuration.OnValueChanged(CCVars.DiscordPrefix, OnPrefixChanged, true);
+        _client = new DiscordSocketClient(new DiscordSocketConfig()
+        {
+            GatewayIntents = GatewayIntents.All
+        });
+        _client.Log += Log;
+        _client.MessageReceived += OnCommandReceivedInternal;
+        _client.MessageReceived += OnMessageReceivedInternal;
+
+        if (_configuration.GetCVar(CCVars.DiscordToken) is not { } token || token == string.Empty)
+        {
+            _sawmill.Info("No Discord token specified, not connecting.");
+            // The Bot is not connected, so we need to set the client to null, because some methods check if the bot is connected using a null check on the client.
+            _client = null;
```

Comment:
You should check this before creating `_client`, so that you don't have to throw it away later.